### PR TITLE
feat(ptv3): export serialized pooling metadata inputs

### PR DESCRIPTION
**Pull request by mojomex**
## Summary
- Port AWML PR 206 serialized-pooling export contract to Autoware-ML PTv3 segmentation and detection exports.
- Add per-stage serialized-pooling metadata inputs to the shared PTv3 backbone export path.
- Generate PTv3 ONNX input names and dynamic axes from the same export specs, instead of duplicating stage-count-sensitive metadata in YAML.
- Update PTv3 deploy configs and docs to describe the ONNX preprocessing contract.
- Add tests for export-mode pooling metadata, generated dynamic axes, and expanded PTv3 detection export input names.

## Stack
- Stacked on #47, which adds the generic `ExportSpec.dynamic_axes` hook used here.

## Checkpoint compatibility
- Autoware-ML-trained checkpoint from `/home/maxschmeller/Downloads/voxel012_122m_t4dataset_j6gen2.tar.xz` loads directly with full coverage: 272 keys loaded, 0 unexpected keys.
- Older AWML `.pth` payloads are structurally compatible (`state_dict` payload), but may need architecture/key-name alignment. The tested AWML `ptv3_t18/model_best.pth` needs `model.backbone.mlp_ratio=2` and legacy key conversion (`seg_head.*` -> `seg3d_head.*`, `backbone.embedding.stem.conv/norm.*` -> `backbone.embedding.stem.0/1.*`).

## Re-export verification
- Canonical `autoware-ml deploy` reaches CUDA/config setup, but cannot complete on this host because `AUTOWARE_ML_DATA_PATH` is unset for the T4Dataset predict dataloader.
- Re-exported the Autoware-ML-trained `best.ckpt` with a synthetic PTv3 batch through the same model/export code path.
- Produced `/tmp/autoware_ml_voxel012_reexport_spec_axes/backbone.onnx` and `/tmp/autoware_ml_voxel012_reexport_spec_axes/seg3d_head.onnx`.
- `onnx.checker.check_model` passed for both.
- Re-exported backbone has 31 inputs; the packaged artifact backbone had the old 3-input interface.
- Inspected ONNX symbolic dims: metadata axes are present, including `serialized_pooling_0_out_voxels_plus_one` for `indptr`, axis-1 `serialized_pooling_0_out_voxels` for serialized order/inverse, and `num_voxels` for decoded point outputs.

## Tests
- `pixi run --environment dev pytest autoware_ml/tests/utils/test_deploy.py autoware_ml/tests/models/test_ptv3_backbone.py autoware_ml/tests/framework/test_ptv3_detection_export.py -q`
- `pixi run --environment dev pre-commit run --files autoware_ml/models/segmentation3d/ptv3_base.py autoware_ml/models/segmentation3d/ptv3.py autoware_ml/models/detection3d/ptv3.py autoware_ml/tests/models/test_ptv3_backbone.py autoware_ml/tests/framework/test_ptv3_detection_export.py autoware_ml/configs/tasks/segmentation3d/ptv3/base.yaml autoware_ml/configs/tasks/detection3d/ptv3/base.yaml`

### CODE REVIEW

**Comment by amadeuszsz** [comment]
@codex review

File: autoware_ml/models/segmentation3d/backbones/ptv3.py [hunk 1]
```diff
@@ -102,6 +103,84 @@ class PointModule(nn.Module):
...
+
+
+def _pooling_depth(stride: int) -> int:
+    """Return how many serialized-code bit triplets are removed by one pooling stride."""
+    depth = 0
+    value = int(stride)
+    while value > 1:
+        depth += 1
+        value >>= 1
+    return depth
```
[the remark this answers is not in the document]
Comment (mojomex) [reply autoware_ml/models/segmentation3d/backbones/ptv3.py]: PTv3 (using SerializedPooling) does not allow non-power-of-2 strides. It asserts and says so withing its SerializedPooling class.

File: autoware_ml/models/segmentation3d/backbones/ptv3.py [hunk 2]
```diff
@@ -102,6 +103,84 @@ class PointModule(nn.Module):
...
+
+def build_serialized_pooling_meta(
+    grid_coord: torch.Tensor,
+    serialized_code: torch.Tensor,
+    serialized_order: torch.Tensor,
+    stride: int,
+) -> tuple[SerializedPoolingMeta, torch.Tensor]:
+    """Build ONNX-facing serialized-pooling metadata for one encoder stage."""
+    depth = _pooling_depth(stride)
+    pooled_code = serialized_code >> (depth * 3)
```
[the remark this answers is not in the document]
Comment (mojomex) [reply autoware_ml/models/segmentation3d/backbones/ptv3.py:135]: I'll have a look whether there's any valid use case where this could happen.

File: autoware_ml/models/segmentation3d/backbones/ptv3.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mojomex) [reply autoware_ml/models/segmentation3d/backbones/ptv3.py]: Update: I checked again — upstream Pointcept does assert power-of-two strides in `SerializedPooling.__init__`, but that assert was dropped in this port, so this was indeed unguarded here. Fixed in cde3efe: `_pooling_depth()` now uses the same `(ceil(stride) - 1).bit_length()` formula as the runtime path and raises `ValueError` on non-power-of-two strides (covering the export helper), and `SerializedPooling.__init__` computes its depth through the same helper, so it raises at construction and `forward` reuses the stored value — the two sites can no longer drift apart. Tests added for both.

File: autoware_ml/models/segmentation3d/backbones/ptv3.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mojomex) [reply autoware_ml/models/segmentation3d/backbones/ptv3.py:135]: Looked into it: the divergence is real but only reachable through the config, not through input data. In export mode the serialization depth is baked from `point_cloud_range`/`grid_size` (not derived from the scan), so a degenerate scan still carries full-depth codes and preprocessing/pooling agree on every input. The fallback can only matter when the *config-derived* depth is smaller than the cumulative pooling depth — i.e. a point cloud spanning fewer than 2^num_stages voxels, which no usable config does.

Rather than replicating the train-time fallback in metadata generation (which would make the export contract shape-polymorphic for an unusable config), cde3efe adds `validate_serialization_geometry()`, called from `PTv3BaseModel.__init__`: it requires config serialization depth >= total pooling depth and raises at model construction, for both training and export. The train-time data-dependent fallback stays as-is for degenerate scans; given a valid config, the exported graph is self-consistent for any input.

**Review by amadeuszsz — APPROVED** [review]
Tested exported model and it works!
LGTM!
